File: results/0.0/DNS/dns-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <arpa/inet.h>
#include <hammer/hammer.h>
#include <hammer/parsers.h>

typedef struct {
    uint16_t transaction_id;
    struct {
        uint8_t qr:1;
        uint8_t opcode:4;
        uint8_t aa:1;
        uint8_t tc:1;
        uint8_t rd:1;
        uint8_t ra:1;
        uint8_t z:3;
        uint8_t rcode:4;
    } __attribute__((packed)) flags;
    uint16_t qdcount;
    uint16_t ancount;
    uint16_t nscount;
    uint16_t arcount;
} DNSHeader;

typedef struct {
    char* name;
    uint16_t type;
    uint16_t class;
} DNSQuestion;

typedef struct {
    char* name;
    uint16_t type;
    uint16_t class;
    uint32_t ttl;
    uint16_t rdlength;
    uint8_t* rdata;
} DNSResourceRecord;

typedef struct {
    DNSHeader header;
    DNSQuestion* questions;
    DNSResourceRecord* answers;
    DNSResourceRecord* authorities;
    DNSResourceRecord* additionals;
    uint16_t question_count;
    uint16_t answer_count;
    uint16_t authority_count;
    uint16_t additional_count;
} DNSMessage;

HParser* parse_dns_header(void) {
    return h_sequence(
        h_uint16(),   // transaction_id
        h_bits(1, false),  // qr
        h_bits(4, false),  // opcode
        h_bits(1, false),  // aa
        h_bits(1, false),  // tc
        h_bits(1, false),  // rd
        h_bits(1, false),  // ra
        h_bits(3, false),  // z
        h_bits(4, false),  // rcode
        h_uint16(),   // qdcount
        h_uint16(),   // ancount
        h_uint16(),   // nscount
        h_uint16()    // arcount
    );
}

HParser* parse_dns_name(void) {
    return h_many1(
        h_sequence(
            h_length_value(h_uint8(), h_ch_range(0x20, 0x7E)),
            h_end_p()
        )
    );
}

HParser* parse_dns_question(void) {
    return h_sequence(
        parse_dns_name(),   // name
        h_uint16(),         // type
        h_uint16()          // class
    );
}

HParser* parse_dns_resource_record(void) {
    return h_sequence(
        parse_dns_name(),   // name
        h_uint16(),         // type
        h_uint16(),         // class
        h_uint32(),         // ttl
        h_length_value(h_uint16(), h_uint8())  // rdata
    );
}

HParser* parse_dns_message(void) {
    return h_sequence(
        parse_dns_header(),
        h_repeat_n(parse_dns_question(), 1),
        h_repeat_n(parse_dns_resource_record(), 1),
        h_repeat_n(parse_dns_resource_record(), 1),
        h_repeat_n(parse_dns_resource_record(), 1)
    );
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <dns_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    if (fread(buffer, 1, file_size, file) != file_size) {
        perror("File read error");
        free(buffer);
        fclose(file);
        return 1;
    }
    fclose(file);

    HParser* dns_parser = parse_dns_message();
    HParseResult* result = h_parse(dns_parser, buffer, file_size);

    if (result && result->ast) {
        printf("DNS message parsed successfully\n");
    } else {
        printf("DNS message parsing failed\n");
    }

    h_parse_result_free(result);
    h_arena_free(h_parser_arena(dns_parser));
    free(buffer);

    return 0;
}

File: results/0.0/DNS/dns-claude-3-5-haiku-20241022.rs
use nom::{
    bytes::complete::{tag, take},
    combinator::{map, opt},
    multi::{count, many0},
    number::complete::{be_u16, be_u8},
    sequence::{tuple, preceded},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct DNSHeader {
    id: u16,
    flags: DNSFlags,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

#[derive(Debug)]
struct DNSFlags {
    qr: bool,
    opcode: u8,
    aa: bool,
    tc: bool,
    rd: bool,
    ra: bool,
    z: u8,
    rcode: u8,
}

#[derive(Debug)]
struct DNSQuestion {
    qname: Vec<String>,
    qtype: u16,
    qclass: u16,
}

#[derive(Debug)]
struct DNSResourceRecord {
    name: Vec<String>,
    rtype: u16,
    rclass: u16,
    ttl: u32,
    rdlength: u16,
    rdata: Vec<u8>,
}

#[derive(Debug)]
struct DNSPacket {
    header: DNSHeader,
    questions: Vec<DNSQuestion>,
    answers: Vec<DNSResourceRecord>,
    authorities: Vec<DNSResourceRecord>,
    additionals: Vec<DNSResourceRecord>,
}

fn parse_dns_flags(input: &[u8]) -> IResult<&[u8], DNSFlags> {
    map(be_u16, |flags| DNSFlags {
        qr: (flags & 0x8000) != 0,
        opcode: ((flags & 0x7800) >> 11) as u8,
        aa: (flags & 0x0400) != 0,
        tc: (flags & 0x0200) != 0,
        rd: (flags & 0x0100) != 0,
        ra: (flags & 0x0080) != 0,
        z: ((flags & 0x0070) >> 4) as u8,
        rcode: (flags & 0x000F) as u8,
    })(input)
}

fn parse_dns_header(input: &[u8]) -> IResult<&[u8], DNSHeader> {
    map(
        tuple((
            be_u16,
            parse_dns_flags,
            be_u16,
            be_u16,
            be_u16,
            be_u16,
        )),
        |(id, flags, qdcount, ancount, nscount, arcount)| DNSHeader {
            id,
            flags,
            qdcount,
            ancount,
            nscount,
            arcount,
        },
    )(input)
}

fn parse_dns_name(input: &[u8]) -> IResult<&[u8], Vec<String>> {
    let mut names = Vec::new();
    let mut current_input = input;

    loop {
        let (remaining, length) = be_u8(current_input)?;
        if length == 0 {
            break;
        }

        let (next_input, name) = take(length)(remaining)?;
        names.push(String::from_utf8_lossy(name).to_string());
        current_input = next_input;
    }

    Ok((current_input, names))
}

fn parse_dns_question(input: &[u8]) -> IResult<&[u8], DNSQuestion> {
    map(
        tuple((parse_dns_name, be_u16, be_u16)),
        |(qname, qtype, qclass)| DNSQuestion {
            qname,
            qtype,
            qclass,
        },
    )(input)
}

fn parse_dns_resource_record(input: &[u8]) -> IResult<&[u8], DNSResourceRecord> {
    map(
        tuple((
            parse_dns_name,
            be_u16,
            be_u16,
            be_u32,
            be_u16,
            take_rdata,
        )),
        |(name, rtype, rclass, ttl, rdlength, rdata)| DNSResourceRecord {
            name,
            rtype,
            rclass,
            ttl,
            rdlength,
            rdata: rdata.to_vec(),
        },
    )(input)
}

fn take_rdata(input: &[u8]) -> IResult<&[u8], &[u8]> {
    let (remaining, length) = be_u16(input)?;
    take(length)(remaining)
}

fn parse_dns_packet(input: &[u8]) -> IResult<&[u8], DNSPacket> {
    map(
        tuple((
            parse_dns_header,
            count(parse_dns_question, |header: &DNSHeader| header.qdcount as usize),
            count(parse_dns_resource_record, |header: &DNSHeader| header.ancount as usize),
            count(parse_dns_resource_record, |header: &DNSHeader| header.nscount as usize),
            count(parse_dns_resource_record, |header: &DNSHeader| header.arcount as usize),
        )),
        |(header, questions, answers, authorities, additionals)| DNSPacket {
            header,
            questions,
            answers,
            authorities,
            additionals,
        },
    )(input)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <dns_packet_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1]).expect("Unable to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Unable to read file");

    match parse_dns_packet(&buffer) {
        Ok((_, packet)) => {
            println!("Parsed DNS Packet: {:?}", packet);
        }
        Err(e) => {
            eprintln!("Error parsing DNS packet: {:?}", e);
        }
    }
}